src/pedersen.rs

```diff
@@ -677,29 +692,38 @@ impl Secp256k1 {
 	) -> Result<ProofRange, Error> {
 		let n_bits = 64;
 
-		let (extra_data_len, extra_data) = match extra_data {
-				Some(d) => (d.len(), d),
-				None => (0, vec![]),
+		let (extra_data_len, extra_data_ptr) = match extra_data {
+				Some(d) => (d.len(), d.as_ptr()),
+				None => (0, ptr::null()),
```

Comment:
Note that this change invalidates all existing bulletproofs on T3, so I think this should be saved for mainnet (or hardforked in, but that would require having 2 versions rust-secp256k1-zkp in grin)

Reply:
oh, my!  I didn't realize that could invalidate all old bulletproofs, actually it's a bug fix. Let me check~

Reply:
Both using an empty vector or a null pointer result in valid bulletproofs, but the important part is that the generation and validation need to have the same value. This is because the challenge `x` needs to be the same. That is derived from an internal `commit`, which is changed by this in the secp256k1-zkp library:
```
if (extra_commit != NULL) {
        secp256k1_sha256_initialize(&sha256);
        secp256k1_sha256_write(&sha256, commit, 32);
        secp256k1_sha256_write(&sha256, extra_commit, extra_commit_len);
        secp256k1_sha256_finalize(&sha256, commit);
    }
```

Reply:
@jaspervdm  No! that will not invalidate existing bulletproofs. this fix only impact batch verification. the `_ptr` here will removed next `.as_ptr()` in unsafe part. 
So, it's NOT a hard fork in T3.

Reply:
Looking at [build.rs:101](https://github.com/mimblewimble/grin/blob/master/wallet/src/libtx/build.rs#L101), rangeproofs are generated with `extra_data = None`. This would mean that the input to `secp256k1_bulletproof_rangeproof_prove` method changes from an empty vector (previously) to a null pointer (your version). This changes the challenge (as explained above). So newly generated rangeproofs will be different. I tested it locally:
```
Proof with an empty vector: ed22845438f16612...
Proof with NULL: b24b41379ad2caf5...
```
So proofs generated after this change will be different and not verify on the old version and vice versa.

Reply:
Thanks @jaspervdm  for review~   Please let me add a test case to cover cross compare between the old version and new version, to make things clear for us:)